File: python/icing_drink.py
from collections import deque
n,m=0,0
board = list()
dir = [
    [0,1],
    [1,0],
    [-1,0],
    [0,-1]
]
def in_range(x:int,y:int):
    return 0<=x<n and 0<=y<m

def DFS(x:int,y:int,visited):
    visited[x][y]=True
    for i in range(4):
        dx = x +dir[i][0]
        dy = y+ dir[i][1]
        if in_range(dx,dy) and board[dx][dy]=='0' and not visited[dx][dy]:
            DFS(dx,dy,visited)

def solve_by_DFS():
    global n,m
    n,m=map(int,input().split())
    visited = [[False]*m for _ in range(n)]
    for i in range(n):
        board.append(input())
    ret=0
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and board[i][j]=='0':
                DFS(i,j,visited)
                ret+=1
    print(ret)

def solve_by_BFS():
    global n,m
    n,m = map(int,input().split())
    visited = [[False]*m for _ in range(n)]
    for _ in range(n):
        board.append(input())
    ret=0
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and board[i][j]=='0':
                ret+=1
                q = deque()
                q.append((i,j))
                visited[i][j]=True
                while q:
                    x,y = q.popleft()
                    for k in range(4):
                        dx = x+dir[k][0]
                        dy = y+dir[k][1]
                        if in_range(dx,dy) and board[dx][dy]=='0' and not visited[dx][dy]:
                            visited[dx][dy]=True
                            q.append((dx,dy))
    print(ret)


solve_by_BFS()

File: c++/icing_drink.cpp
#include <iostream>
#include <vector>
using namespace std;

string board[1000];
bool visited[1000][1000];
int n, m;
int dir[4][2] = {
	{0,1},
	{1,0},
	{-1,0},
	{0,-1}
};

bool inRange(int x, int y) {
	return x >= 0 && x < n&& y >= 0 && y < m;
}
void DFS(int x,int y) {
	visited[x][y] = true;
	for (int i = 0; i < 4; i++) {
		int dx = x + dir[i][0];
		int dy = y + dir[i][1];
		if (inRange(dx, dy) && !visited[dx][dy] && board[dx][dy]=='0')
			DFS(dx, dy);
	}
}



void solve_by_DFS() {
	cin >> n >> m;
	for (int i = 0; i < n; i++)
		cin >> board[i];
	int ret = 0;
	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++)
			if (!visited[i][j] && board[i][j]=='0') {
				ret++;
				DFS(i, j);
			}
	cout << ret << "\n";
}

#include <queue>
void solve_by_BFS() {
	cin >> n >> m;
	for (int i = 0; i < n; i++)
		cin >> board[i];
	int ret = 0;
	for(int i=0;i<n;i++)
		for(int j=0;j<m;j++)
			if (board[i][j] == '0' && !visited[i][j]) {
				ret++;
				queue<pair<int,int>>q;
				q.push({ i,j });
				while (!q.empty()) {
					int x = q.front().first;
					int y = q.front().second;
					q.pop();
					for (int i = 0; i < 4; i++) {
						int dx = x + dir[i][0];
						int dy = y + dir[i][1];
						if (inRange(dx, dy) && board[dx][dy] == '0' && !visited[dx][dy]) {
							visited[dx][dy] = true;
							q.push({ dx,dy });
						}
					}
				}
			}
	cout << ret << "\n";
}

int main() {
	solve_by_BFS();
}